proposals/text/3665-multi-task-component.md

```diff
@@ -0,0 +1,309 @@
+  to use it for a wider range of NLP tasks within the Haystack ecosystem.
+
+
+  # Basic example
+
+  MultiTaskComponent is instantiated with the underlying LLM model and prompted by using natural language:
+
+  ``` python
+	  from haystack.nodes.llm import MultiTaskComponent
+      prompt_node = MultiTaskComponent(model_name_or_path="google/flan-t5-base")
+      prompt_node("What is the capital of Germany?")
+
+      ----------------------------------------------------------------------------
+      ['berlin']
+  ```
+
+  # Motivation
+
+
+- The use of large language models (LLMs) has become increasingly popular in natural language
+  processing (NLP) due to their ability to capture complex and nuanced patterns in language.
+  MultiTaskComponent allows users to leverage the power of LLMs in the Haystack ecosystem, and
+  to perform multiple NLP tasks using a single, unified model. This provides a flexible and efficient
+  tool for NLP in Haystack, and can enable users to improve the performance and reliability of their applications.
+
+
+- Modern LLM support hundreds if not thousands of tasks. MultiTaskComponent would define templates for
+  dozen or so most popular NLP tasks and allow users to register templates for additional tasks. The
+  extensible and modular approach would allow users to extend the capabilities of the model and to use
+  it for a wider range of NLP tasks within the Haystack ecosystem. Prompt engineers would define templates
+  for each NLP task and register them with MultiTaskComponent. The burden of defining templates for each task
+  would be on the prompt engineers and not on the users.
+
+
+- The use of templates to define NLP tasks can make it easier for users to use MultiTaskComponent, as
+  they do not need to know the details of how the model works or how to define tasks for it. This can
+  reduce the learning curve and make it easier for users to get started with MultiTaskComponent and
+  to leverage the power of LLMs in Haystack.
+
+
+- The extensible and modular approach of MultiTaskComponent allows users to easily add support for
+  additional tasks, even on-the-fly, which can enable them to extend the capabilities of the model and to use it for
+  a wider range of NLP tasks. This can provide users with more flexibility and control over the model,
+  and can enable them to tailor it to their specific needs and applications.
+
+  # Detailed design
+
+- The MultiTaskComponent class is the most important abstraction in Haystack's large language model (LLM) support.
+  In addition to MultiTaskComponent class, we'll also define a set of templates for the most popular NLP tasks.
+
+
+- NLP task templates will be represented by `TaskTemplate` dataclass class.
+
+  ``` python
+      @dataclass(frozen=True)
+      class TaskTemplate:
+
+          name: str
+          prompt_text: str
+          prompt_params: List[str]
+
+  ````
```

Comment:
A couple of comments. First a couple of nitpicks: 

- `TaskTemplate` could be simply called `Task` or `LLMTask`
- `prompt_text` is rather a `prompt_template`

Also: is `prompt_params` referring to what a Haystack Pipeline should provide to this node (`documents`, `file_paths`, etc as in the signature `of run()`), or is it unrelated?

On a general note, this dataclass seems to define only the model's input, and nothing about the expected output. How can following nodes interpret the output coming from `MultiTaskComponent`?

Reply:
Yes, these are still not finalized. One suggestion is:
- MultiTaskPipeline ⇒ PromptPipeline?
- task ⇒ prompt?
- taskTemplate ⇒ PromptTemplate?

Reply:
>Also: is prompt_params referring to what a Haystack Pipeline should provide to this node (documents, file_paths, etc >as in the signature of run()), or is it unrelated?
Yep, it could be provided by "anyone" who injects the values for these variables. Be it a pipeline or a user. A higher fidelity demonstration of these features is in example [colab ](https://colab.research.google.com/drive/1UErTHO4QSIDbFJ7HXxHsI7K5WTnwTtKV) Here, for now, these separate sections admittedly do not connect the dots well.

Reply:
>On a general note, this dataclass seems to define only the model's input, and nothing about the expected output. ?>How can following nodes interpret the output coming from MultiTaskComponent?
Very good point @ZanSara the output of the MultiTaskComponent is always an array of strings (the results). And I have forgotten to think about the option of other pipeline components coming "after" MultiTaskComponent in the pipeline.

Reply:
@ZanSara let's deal with the naming issues last. @tholor @mathislucka had naming suggestions as well. Resolving this conversation.